source/slang/slang-ir-lower-tuple-types.cpp

```diff
@@ -0,0 +1,197 @@
+// slang-ir-lower-tuple-types.cpp
+
+#include "slang-ir-lower-tuple-types.h"
+#include "slang-ir.h"
+#include "slang-ir-insts.h"
+
+namespace Slang
+{
+    struct TupleLoweringContext
+    {
+        IRModule* module;
+        DiagnosticSink* sink;
+
+        SharedIRBuilder sharedBuilderStorage;
+
+        List<IRInst*> workList;
+        HashSet<IRInst*> workListSet;
+
+        struct LoweredTupleInfo : public RefObject
+        {
+            IRType* tupleType;
+            IRStructType* structType;
+            List<IRStructField*> fields;
+        };
+        Dictionary<IRInst*, RefPtr<LoweredTupleInfo>> mapLoweredStructToTupleInfo;
+        Dictionary<IRInst*, RefPtr<LoweredTupleInfo>> loweredTuples;
+
+        IRType* maybeLowerTupleType(IRBuilder* builder, IRType* type)
+        {
+            if (auto info = getLoweredTupleType(builder, type))
+                return info->structType;
+            else
+                return type;
+        }
+
+        void appendTypeName(StringBuilder& sb, IRInst* inst)
+        {
+            if (auto name = inst->findDecoration<IRNameHintDecoration>())
+            {
+                sb << name->getName();
+            }
+        }
+
+        LoweredTupleInfo* getLoweredTupleType(IRBuilder* builder, IRInst* type)
+        {
+            if (auto loweredInfo = loweredTuples.TryGetValue(type))
+                return loweredInfo->Ptr();
+            if (auto loweredInfo = mapLoweredStructToTupleInfo.TryGetValue(type))
+                return loweredInfo->Ptr();
+
+            if (!type)
+                return nullptr;
+            if (type->op != kIROp_TupleType)
+                return nullptr;
+
+            RefPtr<LoweredTupleInfo> info = new LoweredTupleInfo();
+            info->tupleType = (IRType*)type;
+            auto structType = builder->createStructType();
+            info->structType = structType;
+            StringBuilder nameSb, fieldNameSb;
+            nameSb << "Tuple";
+            for (UInt i = 0; i < type->getOperandCount(); i++)
+            {
+                auto elementType = maybeLowerTupleType(builder, (IRType*)(type->getOperand(i)));
+                nameSb << "_";
+                appendTypeName(nameSb, elementType);
+                auto key = builder->createStructKey();
+                fieldNameSb.Clear();
+                fieldNameSb << "value" << i;
+                builder->addNameHintDecoration(key, fieldNameSb.getUnownedSlice());
+                auto field = builder->createStructField(structType, key, (IRType*)elementType);
+                info->fields.add(field);
+            }
+            builder->addNameHintDecoration(structType, nameSb.getUnownedSlice());
```

Comment:
While I appreciate the work that has been put into giving the generated tuple types decent names, it might be easier and more consistent with the direction of the IR to give them all the same name hint of `Tuple` and then let the emit logic turn those into `Tuple_0`, `Tuple_1`, etc.

Reply:
Done.